src/rust/ide/lib/view/src/graph_editor/component/edge.rs

```diff
@@ -39,118 +28,550 @@ const LINE_SHAPE_WIDTH   : f32 = LINE_WIDTH + 2.0 * PADDING;
+/// Joint definition
+pub mod joint {
+    use super::*;
+
+    ensogl::define_shape_system! {
+            (
+            ) {
+                let radius : Var<Pixels> = "min(input_size.y,input_size.x)".into();
+                let joint                   = Circle((radius-PADDING.px())/2.0);
+                let joint_color:color::Rgba = color::Lcha::new(0.6,0.5,0.76,1.0).into();
+                let joint_colored           = joint.fill(joint_color);
+                joint_colored.into()
+            }
+        }
+}
+
+fn corner_base_shape
+(radius:&Var<f32>, width:&Var<Pixels>, angle:&Var<f32>, start_angle:&Var<f32>) -> AnyShape {
+    let radius         = 1.px() * radius;
+    let width2         = width / 2.0;
+    let radius_outer   = &radius + &width2;
+    let radius_inner   = &radius - &width2;
+    let ring           = Circle(radius_outer) - Circle(radius_inner);
+    let right:Var<f32> = RIGHT_ANGLE.into();
+    let rot            = right - angle/2.0 + start_angle;
+    let mask           = Plane().cut_angle_fast(angle.clone()).rotate(rot);
+    let shape          = ring * mask;
+    shape.into()
+}
+
+// FIXME [WD]: The 2 follwoing impls are almost the same. Should be merged. This task should will
+//             handled by Wojciech.
+macro_rules! define_corner_start {($color:expr, $highlight_color:expr) => {
     /// Shape definition.
     pub mod corner {
         use super::*;
+
         ensogl::define_shape_system! {
-            (radius:f32, angle:f32, start_angle:f32, pos:Vector2<f32>, dim:Vector2<f32>) {
-                let radius = 1.px() * radius;
-                let width  = LINE_WIDTH.px();
-                let width2 = width / 2.0;
-                let ring   = Circle(&radius + &width2) - Circle(radius-width2);
-                let right : Var<f32> = (RIGHT_ANGLE).into();
-                let rot    = right - &angle/2.0 + start_angle;
-                let mask   = Plane().cut_angle_fast(angle).rotate(rot);
-                let shape  = ring * mask;
+            ( radius:f32
+            , angle:f32
+            , start_angle:f32
+            , pos:Vector2<f32>
+            , dim:Vector2<f32>
+            , focus_split_center:Vector2<f32>
+            , focus_split_angle:f32
+            ) {
+                let width  = &LINE_WIDTH.px();
+                let shape  = corner_base_shape(&radius,width,&angle,&start_angle);
 
                 let shadow_size = 10.px();
-                let n_radius = &shadow_size + 1.px() * dim.y();
-                let n_shape  = Rect((&shadow_size*2.0 + 2.px() * dim.x(),&n_radius*2.0)).corners_radius(n_radius);
-                let n_shape  = n_shape.fill(color::Rgba::new(1.0,0.0,0.0,1.0));
+                let node_radius = &shadow_size + 1.px() * dim.y();
+                let node_shape  = Rect(
+                    (&shadow_size*2.0 + 2.px() * dim.x(),&node_radius*2.0)).corners_radius(node_radius);
+                let node_shape  = node_shape.fill(color::Rgba::new(1.0,0.0,0.0,1.0));
```

Comment:
again, this is refactored somewhere to constants. Lets replace all such colors by the constant.